common/ingressgateway/ingressgateway.go

```diff
@@ -0,0 +1,98 @@
+package ingressgateway
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	// This environmental variable must hold FQDN of ingress gateway which client should access
+	ServiceNameEnv = "INGRESSGATEWAY_FQDN"
+)
+
+// Dialer is an interface which abstracts net.Dialer.
+type Dialer interface {
+	DialContext(ctx context.Context, network, address string) (net.Conn, error)
+}
+
+// clientCreator abstracts interactions with outside world for easy testing.
+type clientCreator struct {
+	Getenv        func(string) string
+	LookupHost    func(string) ([]string, error)
+	GetMinikubeIP func() (string, error)
+	Dialer        Dialer
+}
+
+// Default returns clientCreator ready for production use.
+func Default() *clientCreator {
```

Comment:
Consider adding second "constructor" here, or make clientCreator exported.
This implementation forces users of this library to execute "minikube ip" as a child process. Some clients might want to handle it differently (having this value in an env variable, for example)

Reply:
Hmm, not sure about that. Those functions are abstracted only to make it possible to test the stuff. If it wasn't for testing those functions wouldn't be even configurable. I'd like to have one more opinion before changing that.

Reply:
I will hide clientCreator behind interface so it can be passed around, but its fields are not public. Also the name INGRESSGATEWAY_FQDN is not reflecting the possibility to pass IP here, so I'll change it to INGRESSGATEWAY_ADDRESS. Thanks to that client can overcome dependency on minikube by passing `INGRESSGATEWAY_ADDRESS={minikube ip}`.